Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -788,32 +562,14 @@ - (void)viewWillAppear:(BOOL)animated
 
 - (NSIndexPath *_Nullable)indexPathOfUnreadMessagesIndicator
 {
-    __block TSUnreadIndicatorInteraction *_Nullable unreadIndicator = nil;
-    [self.uiDatabaseConnection readWithBlock:^(YapDatabaseReadTransaction *transaction) {
-        [[TSDatabaseView threadSpecialMessagesDatabaseView:transaction]
-            enumerateRowsInGroup:self.thread.uniqueId
-                      usingBlock:^(
-                          NSString *collection, NSString *key, id object, id metadata, NSUInteger index, BOOL *stop) {
-
-                          if ([object isKindOfClass:[TSUnreadIndicatorInteraction class]]) {
-                              unreadIndicator = (TSUnreadIndicatorInteraction *)object;
-                              *stop = YES;
-                          }
-                      }];
-    }];
-
-    if (!unreadIndicator) {
-        return nil;
-    }
-
-    // TODO: We could do binary search.
-    int numberOfMessages = (int)[self.messageMappings numberOfItemsInGroup:self.thread.uniqueId];
-    for (int i = 0; i < numberOfMessages; i++) {
-        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:i inSection:0];
-        id<OWSMessageData> message = [self messageAtIndexPath:indexPath];
-        if (message.messageType == TSUnreadIndicatorAdapter) {
-            return indexPath;
+    NSInteger row = 0;
+    for (ConversationViewItem *viewItem in self.viewItems) {
```

Comment:
Instead of a lazy cache of "adapters", we now keep around a list/map of "view items", which are view models for each cell in the load window.